src/plugins/vis_builder/public/visualizations/vega/utils/helpers.ts

```diff
@@ -79,30 +79,93 @@ const flattenSeries = (
   });
 };
 
+export interface FlattenedSliceItem {
+  [key: string]: any;
+  value: number;
+  split?: string;
+}
+
+export interface FlattenHierarchyResult {
+  flattenedData: FlattenedSliceItem[];
+  levels: string[];
+}
+
+/**
+ * Flattens hierarchical slice data into a single array of data points
+ * @param {any} data - The hierarchical data to flatten
+ * @param {any[]} group - The group data (rows or columns) if split dimensions exist
+ * @returns {FlattenedSliceItem[]} Flattened array of data points
+ */
+const flattenHierarchy = (data, group): FlattenHierarchyResult => {
+  const flattenedData: FlattenedSliceItem[] = [];
+  const levelSet = new Set<string>();
+
+  const flattenSlices = (
+    slices: any,
+    split?: string,
+    level = 1,
+    parentLabels: { [key: string]: string } = {}
+  ) => {
+    slices.children.forEach((child: any) => {
+      const currentLabels = { ...parentLabels, [`level${level}`]: child.name };
+      levelSet.add(`level${level}`);
+
+      if (child.children && child.children.length > 0) {
+        flattenSlices(child, split, level + 1, currentLabels);
+      } else {
```

Comment:
Is this `else` safe to happen when `child.children = []`?

Reply:
This part of the code is generally safe to execute. Every leaf node, nodes without children, has a size property. However, we could make it safer. I have updated in the code.